newrelic/api/opentelemetry.py

```diff
@@ -0,0 +1,490 @@
+from newrelic.api.time_trace import current_trace, notice_error
+from newrelic.api.transaction import Sentinel, current_transaction
+from newrelic.api.web_transaction import WebTransaction
+from newrelic.core.otlp_utils import create_resource
+
+_logger = logging.getLogger(__name__)
+
+# Attributes that help distinguish span types are
+# sometimes added after span creation and sometimes
+# they are added during span creation.  During the case
+# of the former, we have no way of knowing what sort of
+# library instrumentation to apply, so we have a
+# dictionary of module types as a safeguard.
+#
+# Transactions:
+#   "web": Web Application
+#   "wsgi": WSGI Web Application
+#   "asgi": ASGI Web Application
+#   "other": Background Task
+#   "message": Message Queue
+# Traces:
+#   "other": Function Trace
+#   "db": Database/Datastore
+#   "message": Message Queue
+#
+INSTRUMENTING_MODULE_TYPE = {
+    "Redis": "db",
+    "Mongodb": "db",
+    "Dynamodb": "db",
+    "Pika": "message",
+    "Kombu": "message",
+    "Kafka": "message",
+    "Flask": "wsgi",
+    "Requests": "external",
+}
+# ----------------------------------------------
+# Custom OTel Spans and Traces
+# ----------------------------------------------
+
+# TracerProvider: we can think of this as the agent instance.  Only one can exist
+# SpanProcessor: we can think of this as an application.  In NR, we can have multiple applications
+#   though right now, we can only do SpanProcessor and SynchronousMultiSpanProcessor
+# Tracer: we can think of this as the transaction.
+# Span: we can think of this as the trace.
+# Links functionality has now been enabled but not implemented yet.  Links are relationships
+#   between spans, but lateral in hierarchy.  In NR we only have parent-child relationships.
+#   We may want to preserve this information with a custom attribute.  We can also add this
+#   as a new attribute in a trace, but it will still not be seen in the UI other than a trace
+#   attribute.
+
+
+class Span(otel_api_trace.Span):
+    def __init__(
+        self,
+        name=None,
+        parent=None,  # SpanContext
+        resource=None,
+        attributes=None,
+        kind=otel_api_trace.SpanKind.INTERNAL,
+        nr_transaction=None,
+        nr_trace_type=FunctionTrace,
+        instrumenting_module=None,
+        *args,
+        **kwargs,
+    ):
+        self.name = name
+        self.otel_parent = parent
+        self.attributes = attributes or {}
+        self.kind = kind
+        self.nr_transaction = (
+            nr_transaction or current_transaction()
+        )  # This attribute is purely to prevent garbage collection
+        self.nr_trace = None
+        self.instrumenting_module = instrumenting_module
+
+        self.nr_parent = None
+        current_nr_trace = current_trace()
+        if (
+            not self.otel_parent
+            or (self.otel_parent and self.otel_parent.span_id == int(current_nr_trace.guid, 16))
```

Comment:
Is the casting necessary here?
Could the current_nr_trace come back as None and cause a crash here?